components/support/sql/src/open_database.rs

```diff
@@ -87,21 +89,70 @@ pub fn open_database_with_flags<CI: ConnectionInitializer, P: AsRef<Path>>(
     path: P,
     open_flags: OpenFlags,
     connection_initializer: &CI,
+) -> Result<Connection> {
+    let oi = OpenInfo {
+        path: path.as_ref(),
+        open_flags,
+        connection_initializer,
+        handle_fatal_errors: true,
+    };
+    do_open_database_with_flags(oi)
+}
+
+// This structure helps us with our error recovery strategy.
+struct OpenInfo<'a, CI: ConnectionInitializer> {
+    path: &'a Path,
+    open_flags: OpenFlags,
+    connection_initializer: &'a CI,
+    handle_fatal_errors: bool,
+}
+
+fn do_open_database_with_flags<CI: ConnectionInitializer>(
+    open_info: OpenInfo<'_, CI>,
 ) -> Result<Connection> {
     // Try running the migration logic with an existing file
     log::debug!("{}: opening database", CI::NAME);
-    let mut conn = Connection::open_with_flags(path, open_flags)?;
-    let run_init = should_init(&conn)?;
+    // It turns out Connection::open_with_flags appears to always return without checking
+    // the validity - but that might not be true in all cases, so be safe.
+    let mut conn = match Connection::open_with_flags(open_info.path, open_info.open_flags) {
+        Ok(conn) => conn,
+        Err(e) => {
+            return handle_db_failure(open_info, DroppableConnection::None, Error::SqlError(e))
+        }
+    };
+    log::debug!("{}: checking if initialization is necessary", CI::NAME);
+    let run_init = match should_init(&conn) {
+        Ok(b) => b,
+        Err(e) => return handle_db_failure(open_info, DroppableConnection::Connection(conn), e),
+    };
 
     log::debug!("{}: preparing", CI::NAME);
-    connection_initializer.prepare(&conn)?;
+    if let Err(e) = open_info.connection_initializer.prepare(&conn) {
+        return handle_db_failure(open_info, DroppableConnection::Connection(conn), e);
+    }
 
-    if open_flags.contains(OpenFlags::SQLITE_OPEN_READ_WRITE) {
-        let tx = conn.transaction()?;
+    if open_info
+        .open_flags
+        .contains(OpenFlags::SQLITE_OPEN_READ_WRITE)
+    {
+        let tx = match conn.transaction() {
+            Ok(t) => t,
+            Err(e) => {
+                return handle_db_failure(
+                    open_info,
+                    // Hrm - can't pass 'conn' here, nor tx?!?
```

Comment:
It sucks that here we've lost the ability to pass `conn`, but don't yet have the ability to pass `tx`